sklearn/cluster/tests/test_kernel_kmeans.py

```diff
@@ -0,0 +1,107 @@
+""" Testing for Kernel K means"""
+import sys
+
+import numpy as np
+
+from sklearn.utils.testing import assert_equal
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import assert_raises
+from sklearn.utils.testing import assert_raises_regex
+
+from sklearn.cluster import KernelKMeans
+from sklearn.datasets.samples_generator import make_blobs
+from sklearn.externals.six.moves import cStringIO as StringIO
+
+centers = np.array([
+    [0.0, 5.0, 0.0, 0.0, 0.0],
+    [1.0, 1.0, 4.0, 0.0, 0.0],
+    [1.0, 0.0, 0.0, 5.0, 1.0],
+])
+n_samples = 100
+n_clusters, n_features = centers.shape
+X, true_labels = make_blobs(n_samples=n_samples, centers=centers,
+                            cluster_std=1., random_state=42)
+
+
+def test_kernel_kmeans_dtype():
+    rnd = np.random.RandomState(0)
+    X = rnd.normal(size=(40, 2))
+    X = (X * 10).astype(np.uint8)
+    clf = KernelKMeans().fit(X)
+    pred_x = clf.predict(X)
+    assert_array_equal(clf.labels_, pred_x)
+
+
+def _check_fitted_model(clf):
+    labels = clf.labels_
+    assert_equal(np.unique(labels).shape[0], n_clusters)
+
+    # check error on dataset being too small
+    assert_raises(ValueError, clf.fit, [[0., 1.]])
+
+
+def test_predict():
+    clf = KernelKMeans(n_clusters=n_clusters, random_state=42)
+    clf.fit(X)
+
+    # sanity check: re-predict labeling for training set samples
+    pred = clf.predict(X)
+    assert_array_equal(pred, clf.labels_)
+
+    # re-predict labels for training set using fit_predict
+    pred = clf.fit_predict(X)
+    assert_array_equal(pred, clf.labels_)
+
+
+def test_linear_kernel():
+    clf = KernelKMeans(n_clusters=n_clusters, kernel='linear',
+                       random_state=42).fit(X)
+    _check_fitted_model(clf)
+
+
+def test_polynomial_kernel():
+    clf = KernelKMeans(n_clusters=n_clusters, kernel='polynomial', degree=3,
+                       coef0=2, gamma=0.5, random_state=42).fit(X)
+    _check_fitted_model(clf)
+
+
+def test_sigmoid_kernel():
+    clf = KernelKMeans(n_clusters=n_clusters, kernel='sigmoid', gamma=0.5,
+                       coef0=2, random_state=42).fit(X)
+    _check_fitted_model(clf)
+
+
+def test_rbf_kernel():
+    clf = KernelKMeans(n_clusters=n_clusters, kernel='rbf', gamma=0.5,
+                       random_state=42).fit(X)
+    _check_fitted_model(clf)
```

Comment:
It would be clearer to use a test generator for this – e.g.

```python
def test_kernels():
    def check_kernel(kernel):
        clf = KernelKMeans(n_clusters=n_clusters, kernel='rbf', gamma=0.5,
                           random_state=42)
        clf.fit(X)
        _check_fitted_model(clf)

    for kernel in [None, 'linear', 'rbf', 'polynomial', 'sigmoid']:
        yield check_kernel, kernel
```
for less indirection, you might even put the ``check_fitted_model`` logic in the same function.